Hibernate with Panache: rework getEntityManager and flush

loicmathieu opened the pull request:
Fixes #15595
Fixes #14542

loicmathieu committed:
feat: rework getEntityManager and flush

loicmathieu commented:
@FroMage before implementing the Kotlin and reactive side I wanted to have your feedback on this one:
- I remove the `getEntityManager(Class<?>)` method from PanacheRepository as PanacheEntity and PanacheRepository should only give access to the entity manager of the current entity
- I had to move the `Panache` class, but as it's on the same package it should be fine, but I would need to duplicate it for Kotlin users :(
- I'm not fan of the methods on `Panache` that takes the class to retrieve the right entity manager, the class is used to retrive the name of the persistance unit, maybe those methods should take a string as parameter ?

FroMage commented:
> I remove the getEntityManager(Class<?>) method from PanacheRepository as PanacheEntity and PanacheRepository should only give access to the entity manager of the current entity

+1

> I had to move the Panache class, but as it's on the same package it should be fine, but I would need to duplicate it for Kotlin users :(

Can you explain why you had to move it?

> I'm not fan of the methods on `Panache` that takes the class to retrieve the right entity manager, the class is used to retrive the name of the persistance unit, maybe those methods should take a string as parameter ?

I can see only one: `getEntityManager(Class)` and it makes a lot of sense to me, since that's the only place to be able to obtain an EM for a given class. I do agree we could add a version that takes a `String persistenceUnit`.

In fact, `executeUpdate` and `flush` would need to have some sort of equivalent for specifying an alternate PU, but so far nobody asked for it.

When people ask for it, we can introduce a `PanacheOperations` interface with the same operations as instance methods, and add `PanacheOperations Panache.for(Class/String)` which would return the operations class for the given PU, and `PanacheOperations Panache.default()` for the default PU one that the current static methods point to?

loicmathieu commented:
@FroMage, I added reactive, kotlin and update what's needed in data-rest and spring-data.

> Can you explain why you had to move it?

Because it needs to access the `JpaOperations` that is in the another extension.

> I can see only one: getEntityManager(Class) and it makes a lot of sense to me, since that's the only place to be able to obtain an EM for a given class. I do agree we could add a version that takes a String persistenceUnit.

OK, I keep the `Class` variant and added a `String` variant. Maybe the easiest way will be to replace those two variants with a variant that take the `EntityManager` ? This can also be added to the existing executeUpdate methods without too many overload.

> When people ask for it, we can introduce a PanacheOperations interface 

This is interesting, we can then add more opreations on this `PanacheOperations` (people often ask for some convenience database methods, we can put them here later) if needed and deprecate the one directly on Panache.

FroMage commented:
> OK, I keep the Class variant and added a String variant. Maybe the easiest way will be to replace those two variants with a variant that take the EntityManager ? This can also be added to the existing executeUpdate methods without too many overload.

Well the point of those methods is to make it easy to access operations without getting hold of an EM, so I guess I'd rather not.

FroMage reviewed (dismissed):
Changes look good, but we should add tests for the new methods.

FroMage commented:
Those failing tests don't look good though ;)

loicmathieu commented:
@FroMage I add back the getEntityManager(Class) methods on the repository interface and make them deprecated to avoid too many breaking changes. Can you validate the PR ?

@evanchooly can you have a quick look at the kotlin side?

evanchooly reviewed (approved):
In general, it looks good to me.  there appears to be a mix of calling static methods on subtypes which isn't my favorite but it works.

loicmathieu commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:
```diff
@@ -39,7 +51,7 @@ public static EntityManager getEntityManager(Class<?> clazz) {
      */
     public static TransactionManager getTransactionManager() {
-        return AbstractJpaOperations.getTransactionManager();
+        return JpaOperations.getTransactionManager();
```
Yes it's static.
I'll revert the change.

loicmathieu commented:
@FroMage I added tests as you requested and the failing test are in an unrelated module, is it OK for you ? Can we merge it ?

FroMage reviewed (commented):
LGTM, thanks!
